[feat] 유저 기능 연동

kimhm0828 opened the pull request:
## ✅ 연관된 이슈
#25 

## 📝 작업 내용
- [x] 회원가입
- [x] 아이디 중복체크
- [x] 로그인
- [x] 내정보 확인
- [x] 닉네임 변경
- [x] 비밀번호 변경
----
- [x] access/refresh token 로직 처리하는 래핑 함수 추가
- [x] ->를 axios 이용해서 발전시키기

## 🖼️ 스크린샷 (선택)

kimhm0828 committed:
[feat] 회원가입/중복체크/로그인/정보확인 기능 연동

kimhm0828 committed:
[fix] 브랜치 pull 후 충돌 해결

kimhm0828 committed:
[feat] 유저 기능 연동

kimhm0828 commented on src/app/(auth)/signup/page.tsx:
```diff
@@ -33,39 +34,56 @@ export default function SignupPage() {
-      console.warn("Duplicate ID check should call a secure API endpoint.");
+      const available = await checkDuplicateId(id);
+      if(available){
+        setId(id);
+        alert("사용 가능한 아이디입니다.");
+        setError("");
+      }
+      else{
+        setId("");
+        setError("이미 사용 중인 아이디입니다.");
+      }
     }
```
사용 가능한 아이디라고 알림창으로 표시하는 것과
중복된 아이디일 경우 입력 필드를 비우는 것 둘 다 딱히 바꿀 이유가 없다고 생각해서 바꾸지 않았습니다.

kimingzero replied:
나중에 ui를 생각해서 시간이 된다면 alert를 모달창으로 바꾸면 될 거 같아요

kimhm0828 committed:
[feat] access token 내부 저장소에 저장 및 refresh token 쿠키로 요청하도록 변경

kimhm0828 commented:
/gemini review

kimhm0828 commented on src/app/mypage/changename/page.tsx:
```diff
@@ -6,21 +6,26 @@ import Input from "@/components/common/Input"; // 공통 입력 필드를 불러
       router.push("/mypage"); // 저장이 완료되면 마이페이지로 돌아갑니다.
+    }catch (err) {
+      alert("이름 변경에 실패했습니다.");
+      console.error(err);
```
위와 마찬가지의 이유로 알림창이 사용자 경험을 크게 해칠 것 같지 않아 수정하지 않았습니다.

kimhm0828 commented on src/utils/axiosInstance.ts:
```diff
@@ -0,0 +1,93 @@
+        return axiosInstance(originalRequest);
+      } catch (refreshError) {
+        // refresh token도 실패하면 로그인으로
+        window.location.href = "/login";
```
어차피 로그인을 하러 처음으로 돌아가야 하므로 한번 새로고침하는 것도 괜찮다고 생각해서 수정하지 않았습니다.

kimhm0828 committed:
[refactor] 코드 리뷰를 반영해 에러 처리 로직 변경 등 수정

kimingzero commented on src/app/(auth)/login/page.tsx:
```diff
@@ -28,14 +31,14 @@ export default function LoginPage() {
       setRememberId(true);
     }
-  }, []);
+  }, [router]);
```
useEffect의 의존성이 router인 이유가 없음(이라고 지피티가 그러네요 혹시 이거 지우면 라우팅이 안 되나요?)

현재 코드:

```
useEffect(() => {
  if (typeof window === "undefined") return;
  const savedId = localStorage.getItem("rememberedId");
  if (savedId) {
    setId(savedId);
    setRememberId(true);
  }
}, [router]);
```


👉 router가 바뀔 일이 없음
👉 바뀌면 로그인 페이지에서 다시 실행됨 (불필요)

의존성 배열을 빈 배열로 바꾸는 게 정석이야.

```
useEffect(() => {
  if (typeof window === "undefined") return;
  const savedId = localStorage.getItem("rememberedId");
  if (savedId) {
    setId(savedId);
    setRememberId(true);
  }
}, []); 
```

kimingzero commented on src/app/(auth)/signup/page.tsx:
```diff
@@ -33,39 +34,55 @@ export default function SignupPage() {
+      else{
+        alert("이미 사용중인 아이디입니다.");
+        setError("이미 사용 중인 아이디입니다.");
+      }
```
중복 확인(alert)만 다시 alert 사용 중

이건 Gemini feedback 의도에 안 맞음.

현재:

```
if(available){
  alert("사용 가능한 아이디입니다.");
  setError("");
}
else{
  alert("이미 사용중인 아이디입니다.");
  setError("이미 사용 중인 아이디입니다.");
}

```

➡️ 중복 확인도 inline message로 변경하는 게 UX 통일성에 좋음

개선 예시
```
if (available) {
  setError("사용 가능한 아이디입니다.");
} else {
  setError("이미 사용 중인 아이디입니다.");
}
```


UI에서 초록/빨강 스타일을 나누어 주면 UX가 고급스러워짐.

kimingzero commented on src/app/(auth)/signup/page.tsx:
1. Password 안전성 검증 필요 (보안 UX)

현재 비밀번호 길이/형식 검증 없음.

적어도 다음 정도는 넣는 것이 실무에서 요구됨:

최소 8자

숫자/영문 조합

연속 문자/반복 문자 제한(선택)

경고 메시지 추가 추천:
```
if (password.length < 8) {
  setError("비밀번호는 8자리 이상이어야 합니다.");
  return;
}
```

2. Loading state 없음 (중복 요청/버튼 연타 문제) (선택 사항 - 보고 판단해주세요)

중복 클릭 시 API가 여러 번 호출될 수 있음.

개선 예시
```
const [isLoading, setIsLoading] = useState(false);

const handleSignup = async () => {
  if (isLoading) return;
  setIsLoading(true);
  ...
  setIsLoading(false);
};
```

Button에:
```
<Button disabled={!isAllFieldsFilled || isLoading}>
  {isLoading ? "처리 중..." : "회원가입"}
</Button>
```

kimingzero commented on src/app/(auth)/login/page.tsx:
1. 버튼 연속 클릭 방지 (중복 요청)(선택사항 - 보고 판단해서 수정하시면 돼요)

비동기 요청 중 로그인 버튼을 또 누르면 로그인 요청이 2번 들어간다.

아주 흔한 버그라서 다음처럼 isLoading 상태 넣으면 더 안정적이다:
```
const [isLoading, setIsLoading] = useState(false);

const handleLogin = async () => {
  if (isLoading) return;
  setIsLoading(true);
  ...
  setIsLoading(false);
};
```

Button에:
```
<Button variant="primary" onClick={handleLogin} disabled={isLoading}>
  {isLoading ? "로그인 중..." : "로그인"}
</Button>
```

kimingzero commented on src/app/(auth)/signup/signup.ts:
```diff
@@ -0,0 +1,31 @@
+        skipAuth: true,
+        params: { userId },
+      },
+    );
```
문제점 1 — axios.get 옵션 구조가 잘못됨 -> 뒤에 있는 쉼표만 지우시면 될 것 같습니다.

지금 코드 ↓

```
await axiosInstance.get(`/auth/verify`, {
  skipAuth: true,
  params: { userId },
});
```


axiosInstance.get(url, config)는 한 개의 config object만 받아.
즉, 지금은 문제 없음.

그런데 네가 작성한 코드에서는 config를 2개 넣는 실수를 한 줄로 보이면 오해할 수 있음

지금 코드:

```
await axiosInstance.get(`/auth/verify`, {
    skipAuth: true,
    params: { userId },
  },
);
```


마지막 }, 때문에 객체가 끝난 것처럼 보이고,
추가 config가 이어질 것 같은 모양이야 → 타입스크립트 상 오류는 없지만 가독성 나쁨

정확한 형태는 ↓

```
await axiosInstance.get("/auth/verify", {
  skipAuth: true,
  params: { userId },
});
```

kimingzero commented on src/app/(auth)/signup/signup.ts:
signup — 깔끔하지만 에러 처리 방식 통일 필요 (돌아가는데에는 문제가 없어보이지만 코드 개션부분인거같아 보신 뒤 필요없다고 생각하시면 넘기시면 됩니다)
✔ 좋은 점

성공 시 true 반환 → 간단한 성공/실패 로직

skipAuth true 적용

catch에서 throw로 전달해 UI에서 처리 가능

⚠️ 아쉬운 점 1 — 성공 여부를 boolean으로 단순 반환

백엔드에서 signup 성공 시 userId, nickname 같은 정보를 내려줄 수도 있음.

UI에서 써야 할 경우가 많으니, response.data.data 자체를 반환하는 게 더 좋음.

⚠️ 아쉬운 점 2 — catch(e) → e가 unknown 타입인데 그대로 throw

TypeScript에서는 unknown error는
직접 타입을 좁히거나 ApiError로 감싸는 것이 안정적.

✨ 개선된 signup 코드
```
export const signup = async (payload: SignupParams) => {
  try {
    const res = await axiosInstance.post("/users/signup", payload, {
      skipAuth: true,
    });

    return res.data.data; // 성공 데이터 그대로 반환
  } catch (error: any) {
    console.error("회원가입 요청 오류", error);

    if (error.response?.data?.message) {
      throw new Error(error.response.data.message);
    }

    throw new Error("회원가입 중 알 수 없는 오류가 발생했습니다.");
  }
};
```

kimingzero commented on src/app/mypage/changename/page.tsx:
Loading 상태가 없음 (중복 클릭 가능) - 이것도 위에 있는 파일들과 비슷하게 필요한 내용이라 생각하시면 진행하면 될거같아요

사용자가 버튼을 여러 번 누르면 API가 중복 요청될 수 있음.

개선:
```
const [isLoading, setIsLoading] = useState(false);

const handleSubmit = async () => {
  if (isLoading) return;
  setIsLoading(true);
  
  try { ... }
  finally { setIsLoading(false); }
};
```

그리고 Button:
```
<Button onClick={handleSubmit} disabled={!isButtonEnabled || isLoading}>
  {isLoading ? "저장 중..." : "이름 변경하기"}
</Button>
```

kimingzero commented on src/app/mypage/page.tsx:
```diff
@@ -52,7 +53,8 @@ export default function ProfilePage() {
+  const handleLogout = async() => {
+    await axiosInstance.post("/auth/logout");
     router.push("/login"); // 로그아웃 버튼 클릭 시 로그인 화면으로 이동합니다.
   };
```
Logout 시 토큰 제거 로직 누락됨

현재 logout:
```
await axiosInstance.post("/auth/logout");
router.push("/login");
```

하지만

localStorage에 남아 있는 accessToken

refreshToken

userName

state 관리된 local user info

이런 게 그대로 남아있으면 보안상 문제됨.

개선 필수:
```
await axiosInstance.post("/auth/logout");
localStorage.clear();
sessionStorage.clear();
useAuthStore.getState().clearTokens?.();
router.push("/login");
```

kimingzero commented on src/app/mypage/password/changePassword.ts:
newPw 최소 검증(문자 수, 빈 문자열) 프론트에서 필요

API 호출 전에 최소한:

길이가 8자 이상

공백 문자열 아님

currPw !== newPw

정도는 체크하는 게 UX적으로 더 좋음.

물론 이건 UI단에서 체크하는 게 더 적절하지만
함수 내부에서도 방어적 프로그래밍 가능함.

kimingzero commented on src/app/mypage/password/page.tsx:
새 비밀번호 검증(길이, 규칙)이 필요함

Signup에서는 비밀번호 규칙 검증이 있었으나
ChangePassword에는 없음.

실제 보안 UX에서는 다음 최소 조건이 필요함:

8자 이상

기존 비밀번호와 달라야 함

너무 단순한 문자열 방지

예:
```
if (newPassword.length < 8) {
  setError("비밀번호는 8자리 이상이어야 합니다.");
  return;
}

if (currentPassword === newPassword) {
  setError("기존 비밀번호와 다른 비밀번호를 입력해주세요.");
  return;
}
```

이건 실제 서비스라면 반드시 넣어야 하는 로직.

kimingzero commented on src/app/mypage/password/page.tsx:
Loading state가 없음 (중복 클릭 위험)

버튼을 연속 클릭하면 changePassword API가 2~3번 갈 수 있음.

실무에서는 반드시 막아야 함.

🔧 개선:
```
const [isLoading, setIsLoading] = useState(false);

const handleSubmit = async () => {
  if (isLoading) return;
  setIsLoading(true);
  
  ...
  
  setIsLoading(false);
};
```

그리고 버튼:
```
<Button disabled={!isButtonEnabled || isLoading}>
  {isLoading ? "처리 중..." : "비밀번호 변경하기"}
</Button>
```

kimingzero commented on src/lib/hooks/useUserData.ts:
```diff
@@ -8,23 +9,25 @@ export function useUserData() { // 사용자 이름과 보유 포인트를 제
+        setUserName(data.nickname);
+        setAvailablePoints(data.point);
+      }catch(error){
+        if (error instanceof ApiError) {
```
토큰 만료 시 에러 핸들링 없음

현재
```
if (error instanceof ApiError)
```

하지만 토큰 만료(401)이나 refreshToken 만료(419) 같은 경우
로그아웃 처리 or 로그인 페이지로 이동해야 함.

🔧 개선:
```
if (error instanceof ApiError) {
  if (error.status === 401) {
    router.push("/login");
    return;
  }
}
```

또는 axios interceptor에서 관리해도 됨.

kimhm0828 replied:
토큰이 만료되었을 경우 자동으로 axiosIntercepter에서 로그인 페이지로 핸들링하기 때문에 따로 반영하지 않았습니다.
다만 axiosIntercepter에서 로그인 페이지로 이동할 때 내부 저장소에서 토큰을 삭제하는 로직은 추가했습니다.

kimingzero commented on src/utils/apiError.ts:
개선 필요 포인트 (실무 기준 100점 만들기) - 개선할 부분이라 필수는 아닙니다
1) ❗ Error 상속 시 name을 "ApiError"로 명시해야 함

지금은:
```
export class ApiError extends Error {
```

하지만 Error 상속 시 name이 "Error"로 남아 있어서
instanceof ApiError는 잘 작동하지만
로그나 sentry 등에서 에러 이름이 "Error"로 찍히는 문제가 있음.

🔧 해결:
```
this.name = "ApiError";
```
2) ❗ message 필드는 super(message)로 이미 저장되는데 다시 this.message로 재정의하면 안 됨

현재 코드:
```
super(message);
this.message = message;
```

super가 메시지를 이미 저장함
→ 다시 this.message를 동일하게 덮어쓰는 것은 불필요하고 혼란을 초래할 수 있음.

아예 제거해야 함.

3) ❗ Error 객체는 stack trace를 유지해야 함 → super() 이후 this.stack 유지 확인 필요

일부 브라우저에서는 Error 상속 시 stack이 올바르게 전달되지 않을 수 있음.

아래 코드가 있는 게 안정적:
```
if (Error.captureStackTrace) {
  Error.captureStackTrace(this, ApiError);
}
```
4) ❗ code는 status로 명명하는 것이 REST와 더 일관됨

많은 API에서는 응답에 status(HTTP status code)를 사용함.
이는 가독성과 일관성에서 유리함.

5) ❗ prototype 설정을 명확하게 해야 함

TS에서는 Error 상속 시 prototype chain이 끊기는 문제가 있어.

안정적으로 하려면 반드시 추가해줘야 함:
```
Object.setPrototypeOf(this, ApiError.prototype);
```
✨ 실무 100점짜리 개선된 ApiError

아래는 “axios interceptor + 전역 에러 핸들링 + Sentry” 등에 최적화된 완전한 버전이야.
```
export class ApiError extends Error {
  status: number;
  details?: any;

  constructor(status: number, message: string, details?: any) {
    super(message);

    // Set the prototype explicitly (TS error inheritance fix)
    Object.setPrototypeOf(this, ApiError.prototype);

    this.name = "ApiError";
    this.status = status;
    this.details = details;

    // Optional: Ensure proper stack trace in V8 engines
    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, ApiError);
    }
  }
}
```

kimingzero commented on src/utils/tokenStorage.ts:
하지만… 실무 기준으로는 개선이 필요함 (중요한 포인트)

현재 스토어는 매우 기본 형태기 때문에 실제 서비스에서 토큰 관리를 안정적으로 수행하기에는 여러 부분이 부족해.

refreshToken이 없음 → 자동 로그아웃/갱신 불가

실무에서 accessToken만 저장해도 초반엔 되지만,
토큰이 만료되면 갱신이 불가능함.

필수 추가 항목

refreshToken
setRefreshToken
clearAuth (로그아웃 시 전체 삭제)

2) ❗ localStorage 연동이 없음 → 새로고침하면 로그인 풀림 (심각)

지금 구조에서는:

accessToken → 메모리(Zustand 내부)에만 저장됨
페이지 새로고침하면 Zustand 초기화됨 → 로그인 풀림
검색 결과가 아닌 실제 서비스에서는 영속 저장 필요:

localStorage
혹은 cookies (보안적으로 더 안전)
해결 방식 예시:
```
accessToken: typeof window !== "undefined" ? localStorage.getItem("accessToken") : null,
setAccessToken: (token) => {
  set({ accessToken: token });
  if (typeof window !== "undefined") {
    if (token) localStorage.setItem("accessToken", token);
    else localStorage.removeItem("accessToken");
  }
},
```
3) ❗ 전역에서 clearTokens 기능이 필요함

로그아웃, 토큰 만료, 탈퇴 등 여러 곳에서 필요.

예:
```
clearTokens: () => {
  set({ accessToken: null, refreshToken: null });
  localStorage.removeItem("accessToken");
  localStorage.removeItem("refreshToken");
},
```
4) ❗ persist 미들웨어 적용 가능 (가장 실무적)

Zustand의 persist를 사용하면 localStorage 연동도 매우 간단해짐.

매우 추천되는 형태:
```
import { create } from "zustand";
import { persist } from "zustand/middleware";

export const useAuthStore = create(
  persist(
    (set) => ({
      accessToken: null,
      refreshToken: null,
      setAccessToken: (token) => set({ accessToken: token }),
      setRefreshToken: (token) => set({ refreshToken: token }),
      clearAuth: () => set({ accessToken: null, refreshToken: null }),
    }),
    { name: "auth-storage" } // localStorage key
  )
);
```

이러면 다른 코드가 전혀 변하지 않아도 토큰 영속성 문제가 해결됨.

✨ 실무 100% 버전 Zustand Auth Store (완성판)
```
import { create } from "zustand";
import { persist } from "zustand/middleware";

interface AuthState {
  accessToken: string | null;
  refreshToken: string | null;
  setAccessToken: (token: string | null) => void;
  setRefreshToken: (token: string | null) => void;
  clearAuth: () => void;
}

export const useAuthStore = create<AuthState>()(
  persist(
    (set) => ({
      accessToken: null,
      refreshToken: null,

      setAccessToken: (token) =>
        set(() => ({ accessToken: token })),

      setRefreshToken: (token) =>
        set(() => ({ refreshToken: token })),

      clearAuth: () =>
        set(() => ({
          accessToken: null,
          refreshToken: null,
        })),
    }),
    {
      name: "auth-storage", // localStorage key
    }
  )
);
```

kimhm0828 replied:
refresh token은 백엔드 서버에서 쿠키로 내려받는 형태라 서비스에서 따로 관리하지 않습니다.
실제로는 그냥 refresh 엔드포인트로 요청만 보내면 브라우저에서 자동으로 refresh token이 담긴 쿠키를 넣어서 보냅니다.
오히려 localstorage를 사용하는 쪽이 보안상 더 위험하기 때문에 반영하지 않았습니다.

참고: 백엔드 토큰 로직
```java
@PostMapping("/login")
    public ResponseEntity<BaseResponse<?>> login(@RequestBody LoginReqDto loginReqDto) {
        TokenWithCookie tokenWithCookie = authService.login(loginReqDto);
        LoginResDto loginResDto = new LoginResDto(tokenWithCookie.accessToken());
        return ApiResponse.successWithCookie(SuccessCode.OK, loginResDto, tokenWithCookie.cookie());
    }

@PostMapping("/refresh")
    public ResponseEntity<BaseResponse<?>> refresh(@CookieValue("refreshToken") String refreshToken) {
        TokenWithCookie tokenWithCookie = authService.refresh(refreshToken);
        LoginResDto loginResDto = new LoginResDto(tokenWithCookie.accessToken());
        return ApiResponse.successWithCookie(SuccessCode.OK, loginResDto, tokenWithCookie.cookie());
    }
```
위와 같이 로그인 시 access token + refresh token이 담긴 쿠키 형태로 응답을 건네주고
refresh 엔드포인트로 요청을 보낼 때 브라우저가 넣어 보낸 쿠키에서 refresh token 값을 자동으로 인식해 토큰을 재발급합니다.

kimingzero commented on src/utils/axiosInstance.ts:
실무 기준에서 반드시 개선해야 하는 포인트

정상 작동은 되지만, 서비스 안정성과 보안성 측면에서 개선 필요.

1) ❗ refreshToken 갱신에 **axios.post()**를 쓰고 있음 → 문제 가능성 큼

refresh 요청 부분:
```
const refreshResponse = await axios.post(
  `/auth/refresh`,
  {},
  { headers: { "Content-Type": "application/json" }, withCredentials: true }
);
```

이 axios는 "axiosInstance"가 아니기 때문에:

❌ baseURL 적용 안됨

→ 현재 baseURL이 비어 있으므로 상대 경로로 동작하지만
→ 실제 배포 시 refresh 요청이 엉뚱한 origin으로 갈 가능성이 있음.

❌ interceptors 비적용

refresh 요청에는 interceptors를 적용하지 않는 게 맞지만,
적어도 axiosInstance 설정만큼은 반영해야 한다.

즉,

baseURL
withCredentials 기본값
timeout
공통 headers

이런 것들이 반영되지 않는다.

✔ 해결: axiosInstance에서 “skipAuth: true” 옵션을 사용하여 호출
```
const refreshResponse = await axiosInstance.post(
  "/auth/refresh",
  {},
  { skipAuth: true }
);
```

이렇게 하면:

baseURL 사용됨

interceptors는 skipAuth로 인해 Authorization 안 붙음

withCredentials는 true로 유지됨

가장 안전한 방식임.

2) ❗ 프로퍼티 _retry TypeScript 타입 오류 가능성

axios config에는 _retry라는 필드가 기본적으로 존재하지 않음.

실제 프로젝트에서는 다음 처리가 필요함:

(1) global.d.ts에 axios config 확장
```
declare module "axios" {
  export interface AxiosRequestConfig {
    skipAuth?: boolean;
    _retry?: boolean;
  }
}
```

이미 skipAuth는 정의했으니, _retry도 반드시 추가하는 것이 필요함.

3) ❗ refresh 실패 시 logout 로직 부족

현재:
```
window.location.href = "/login";

```
그러나 accessToken과 refreshToken이 localStorage/Zustand에 남아있음.

즉 로그인 화면으로는 갔지만 토큰은 그대로 남아 있음.

다음 시도 때 오작동 가능.

반드시 토큰 삭제 필요:
```
const { clearAuth } = useAuthStore.getState();
clearAuth();
window.location.href = "/login";
```

이 clearAuth는 우리가 앞에서 만든 실무용 Zustand auth store에 포함되어 있어야 함.

4) ❗ skipAuth가 true면 withCredentials = false로 설정하는 게 위험할 수 있음

현재 skipAuth 처리:
```
config.headers.Authorization = undefined;
config.withCredentials = false;
```

문제:

refresh token이 쿠키에 담겨 있다면 withCredentials=false는 쿠키 전송을 차단함

로그인 요청에서 refresh token이 필요할 수는 없지만

다른 skipAuth 요청(회원가입, 중복확인 등)에서 쿠키가 필요한 경우를 차단하게 됨

즉, skipAuth = true는 "토큰이 필요 없는 요청"이지
"쿠키를 보내지 않는 요청"은 아님.

개선:

skipAuth에서는 Authorization만 제거해야 함.

❌ 수정 전
withCredentials = false

✔ 수정 후
withCredentials 그대로 유지

5) ❗ ApiError 생성 시 message, code가 undefined일 수 있음

현재 코드:
```
const { code, message, data } = error.response.data;
return Promise.reject(new ApiError(code, message, data));
```

문제:

백엔드에서 code, message를 응답 안하면 undefined로 들어감

ApiError에 undefined가 들어가면 핸들링이 어려워짐

개선:
```
const backend = error.response.data;
const code = backend.code ?? error.response.status ?? -1;
const message = backend.message ?? "알 수 없는 서버 오류가 발생했습니다.";
```
6) ❗ refresh 요청과 기존 요청 간 Race Condition 방어 없음

두 개 이상의 요청이 동시에 401 발생할 때,

두 개의 refresh 요청이 동시에 날아갈 가능성이 있음

첫 번째는 성공하고, 두 번째는 실패해서 logout 될 수 있음

실무에서는 refresh 요청을 mutex(락)으로 막아야 함.

이건 고급 주제라 필요하면 구현도 해줄 수 있음.

✨ 실무 완성판으로 리팩토링한 코드 (100% 안정 버전)

원하면 위 문제 모두 해결한 완전한 axiosInstance 전체 코드를 만들어줄게.
아래는 간단 개선 버전:
```
import axios from "axios";
import { ApiError } from "./apiError";
import { useAuthStore } from "./tokenStorage";

declare module "axios" {
  export interface AxiosRequestConfig {
    skipAuth?: boolean;
    _retry?: boolean;
  }
}

const axiosInstance = axios.create({
  baseURL: "",
  withCredentials: true,
  headers: { "Content-Type": "application/json" },
});

// 요청 인터셉터
axiosInstance.interceptors.request.use(
  (config) => {
    if (config.skipAuth) {
      delete config.headers.Authorization;
      return config;
    }

    const token = useAuthStore.getState().accessToken;
    if (token) config.headers.Authorization = `Bearer ${token}`;

    return config;
  },
  (error) => Promise.reject(error)
);

// 응답 인터셉터
axiosInstance.interceptors.response.use(
  (res) => res,
  async (error) => {
    const original = error.config;

    const isJwtExpired =
      error.response &&
      (error.response.status === 401 ||
        error.response.data?.code === 40101 ||
        error.response.data?.code === 40102);

    // Refresh Token 처리
    if (isJwtExpired && !original._retry) {
      original._retry = true;

      try {
        const refreshRes = await axiosInstance.post(
          "/auth/refresh",
          {},
          { skipAuth: true }
        );

        const { accessToken } = refreshRes.data.data;
        useAuthStore.getState().setAccessToken(accessToken);

        original.headers.Authorization = `Bearer ${accessToken}`;
        return axiosInstance(original);
      } catch (err) {
        useAuthStore.getState().clearAuth?.();
        window.location.href = "/login";
        return Promise.reject(new ApiError(401, "토큰 갱신 실패"));
      }
    }

    // 일반 에러 처리
    if (error.response) {
      const backend = error.response.data;
      const code = backend.code ?? error.response.status ?? -1;
      const message = backend.message ?? "서버 오류가 발생했습니다.";
      return Promise.reject(new ApiError(code, message, backend.data));
    }

    return Promise.reject(new ApiError(-1, "네트워크 오류가 발생했습니다."));
  }
);

export default axiosInstance;
```

kimingzero reviewed (commented):
마지막 utils 파일 안에 있는 3개의 파일은 지피티를 아무리 써도 이해가 안 되어서 복붙했습니다
전반적으로 다 좋은데 하면 좋은 것들(추천 위주) 이라서 보시고 굳이? 싶은건 코멘트로 굳이? 쓰시고 넘기시면 됩니다

kimhm0828 committed:
[feat] 코드 리뷰 반영

-아이디 검증 로직&안내문구 추가
-비밀번호 검증 로직&안내문구 추가
-중복 요청이 가지 않도록 로딩 로직 추가
-로그아웃 시 내부 저장소에서 토큰 삭제 로직 추가
-refresh token을 이용한 요청에서 race condition 방어: 하나의 요청만 실제로 진행되고 나머지는 promise 객체를 참고하도록

kimhm0828 commented:
/gemini review

kimhm0828 committed:
[refactor] 코드 리뷰 반영, 로직 수정

-로딩 로직 정리 및 간결화
-회원가입의 입력값 검증 로직 정리 및 간결화
-중복 id 검사에서 오류 처리 로직 추가
-그 외 사용하지 않는 코드 / 필요없는 코드 제거

kimhm0828 commented:
/gemini review

kimhm0828 committed:
[fix] 코드리뷰 내용 반영해서 수정

-인자 전달 방식 변경
-apiError 생성할 때 생성자에서 super(message)로 메시지 전달하도록 수정
-axios를 직접 이용하던 코드를 axiosInstance를 이용하도록 수정
-에러 처리 로직 수정
-불필요한 코드 제거
-누락된 코드 추가(로딩 상태 true로 변경)
-alert -> 인라인 에러 메시지로 변경
-유효성 검사 함수들을 validate.ts 파일로 분리해서 관리

kimhm0828 commented:
/gemini review

kimhm0828 committed:
[refactor] 유효성 검사 로직 수정 및 오류 처리 로직 수정

kimhm0828 commented:
/gemini review

kimhm0828 commented on src/app/(auth)/signup/signup.ts:
```diff
@@ -0,0 +1,37 @@
+    return true;
+  }catch(error: unknown) {
+    // 409 Conflict 에러인 경우에만 중복으로 처리
+    if (error instanceof ApiError && error.status === 409) {
```
백엔드에서 409로 응답하고 있어 따로 수정하지 않았습니다.

kimhm0828 commented on src/lib/hooks/useUserData.ts:
```diff
@@ -8,23 +9,25 @@ export function useUserData() { // 사용자 이름과 보유 포인트를 제
 
-    const savedName = localStorage.getItem("userName"); // 저장된 사용자 이름이 있는지 확인합니다.
-    if (savedName) {
-      setUserName(savedName);
+    async function loadUserData() {
+      try{
+        const response = await axiosInstance.get("/users/me");
+        const data = response.data.data;
+
+        setUserName(data.nickname);
+        setAvailablePoints(data.point);
+      }catch(error){
+        if (error instanceof ApiError) {
+          console.error("사용자 정보를 불러오는 중 오류:", error.message);
+        } else {
+          console.error("알 수 없는 오류 발생:", error);
+        }
+      }
     }
 
-    const points = getAvailablePoints(); // 포인트 유틸을 통해 기본 포인트를 계산합니다.
-    setAvailablePoints(points); // 계산된 포인트를 상태에 반영합니다.
+    loadUserData();
   }, []);
```
이 리뷰대로라면 accessToken이 없을 때에는 요청을 보내지 않도록 해야 하는데
accessToken이 없어도 refreshToken을 이용해서 accessToken을 새로 받아오기 위해서는 반드시 axiosInstance를 이용해 요청을 보내야 하기 때문에 반영하지 않았습니다.

kimhm0828 committed:
[fix] 네트워크 에러 처리 로직 추가, 에러 처리 로직 수정

kimingzero commented on src/app/(auth)/login/login.ts:
```diff
@@ -0,0 +1,30 @@
+    return response.data;
+  } catch (error: unknown) {
+    console.error("로그인 요청 오류", error);
+    throw error;
```
로그인 실패 시, ApiError로 변환하지 않음

지금 오류 처리:
```
catch (error) {
    console.error("로그인 요청 오류", error);
    throw error;
}
```

이것의 문제:

axiosInstance에서는 에러를 ApiError로 감싸는데

loginUser에서는 원본 error를 그대로 throw함
→ 즉 ApiError이거나 AxiosError이거나 Error 타입이 여러 종류로 섞임

실제 UI에서 에러 메시지를 일관되게 처리하기 어려움.

👉 실무에서는 일관된 에러 타입으로 throw해야 한다.

🔥 4) id, password 빈 값일 때 서버로 요청 보내버림

프론트에서 validation 처리되어야 함.
```
if (!id || !password) return;
```
🔧 개선된 Best Practice 코드

아래는 실무에서 가장 이상적인 형태:
```
import axiosInstance from "@/utils/axiosInstance";
import { useAuthStore } from "@/utils/tokenStorage";
import { ApiError } from "@/utils/apiError";

interface LoginResponse {
  accessToken: string;
  refreshToken?: string;
}

export async function loginUser(id: string, password: string): Promise<LoginResponse> {
  if (!id.trim() || !password.trim()) {
    throw new ApiError(400, "아이디와 비밀번호를 입력해주세요.");
  }

  try {
    const response = await axiosInstance.post(
      "/auth/login",
      { userId: id, password },
      { skipAuth: true }
    );

    const { accessToken, refreshToken } = response.data.data;

    // Zustand 저장 (accessToken은 Zustand, refreshToken은 cookie일 가능성)
    const { setAccessToken } = useAuthStore.getState();
    setAccessToken(accessToken);

    return { accessToken, refreshToken };
  } catch (error: unknown) {
    // axiosInstance에서 ApiError로 변환해주므로 error는 ApiError 타입
    if (error instanceof ApiError) {
      throw error;
    }
    throw new ApiError(-1, "로그인 요청 중 알 수 없는 오류가 발생했습니다.");
  }
}
```

kimingzero reviewed (dismissed):
merge하셔도 될 거 같아욤

kimhm0828 committed:
[fix] 회원가입 시 에러가 종류에 맞게 정상적으로 보이도록 수정, 에러 처리 수정
